File: Leetcode 30 Days Solution/Minimum Falling Path Sum.cpp
class Solution {
public:
    int minFallingPathSum(vector<vector<int>>& matrix) {
        
        if(matrix.size()==1)
        {
            return matrix[0][0];
        }
        else
        {
            int n=matrix.size();
            int dp[n][n];
            for(int i=0;i<n;i++)
            {
                dp[n-1][i]=matrix[n-1][i];
            }
            int trav=n-2;
            while(trav>=0)
            {
                for(int i=0;i<n;i++)
                {
                    if(i==0)
                    {
                        dp[trav][i]=min(matrix[trav][i]+dp[trav+1][i],matrix[trav][i]+dp[trav+1][i+1]);
                    }
                    else if(i==n-1)
                    {
                        dp[trav][i]=min(matrix[trav][i]+dp[trav+1][i],matrix[trav][i]+dp[trav+1][i-1]);
                    }
                    else
                    {
                        dp[trav][i]=min(matrix[trav][i]+dp[trav+1][i+1],min(matrix[trav][i]+dp[trav+1][i],matrix[trav][i]+dp[trav+1][i-1]));
                    }
                }
                trav--;
            }
            int ans=INT_MAX;
            for(int i=0;i<n;i++)
            {
                ans=min(ans,dp[0][i]);
            }
            return ans;
        }
        
    }
};

File: Python/Minimum Falling Path Sum.py
class Solution:
    def minFallingPathSum(self, matrix: List[List[int]]) -> int:
        
        row=len(matrix)
        col=len(matrix[0])
        listtemp=[[0 for c in range(col)]for r in range(row)]
        for i in range(0,col):
            listtemp[0][i]=matrix[0][i]
        i=1
        while i<row:
            
            for j in range(0,col):
                if j==0 :
                    listtemp[i][j]=min(matrix[i][j]+listtemp[i-1][j],matrix[i][j]+listtemp[i-1][j+1])
                elif j==col-1:
                    listtemp[i][j]=min(matrix[i][j]+listtemp[i-1][j],matrix[i][j]+listtemp[i-1][j-1])
                else:
                    listtemp[i][j]=min(matrix[i][j]+listtemp[i-1][j-1],min(matrix[i][j]+listtemp[i-1][j],matrix[i][j]+listtemp[i-1][j+1]))
            i=i+1
        
        ans=100000
        for i in range(0,col):
            ans=min(listtemp[row-1][i],ans)
        return ans
        
        
        
